### PR TITLE
Updated lodash and refactored plugin hook system

### PR DESCRIPTION
**Pull request by christophwitzko**

### CODE REVIEW

**Comment by janl** [comment]
can you explain the reasoning behind this change.

File: lib/core/hooks.js [hunk 1]
```diff
@@ -34,24 +37,14 @@ module.exports = function (env_config, callback) {
...
-    return _.compact(_.map(env_config.plugins, function(plugin) {
-      if(plugin.hooks[type] && plugin.hooks[type][name]) {
-        return plugin.hooks[type][name];
-      }
-    }));
-  };
-
   var runAsyncEvery = function(type, name, args, callback) {
-    var runHook = function(hook, callback) {
-      args.push(callback);
```
Comment (christophwitzko) [inline lib/core/hooks.js]: Every time `runHook` is called a new `callback` function is pushed to the `args` array. I don't think it should work this way.

File: lib/server/plugins/api/index.js [hunk 2]
```diff
@@ -190,9 +190,10 @@ exports.register = function (plugin, options, next) {
       method: 'GET',
       path: '/_api/_plugins/{name}',
       handler: function (request, reply) {
-        var metaData = _.find(plugins.metadata(options.app), function (doc) {
-          return doc.name === request.params.name;
```
Comment (christophwitzko) [inline lib/server/plugins/api/index.js]: We don't need a compare function. `lodash` can do that for us.

File: lib/core/hooks.js [hunk 3]
```diff
@@ -3,30 +3,33 @@ var async = require('async');
...
-        plugin.hooks[type][name].apply(env_config, args);
-        ran = true;
-      }
+    var hooks = getHooksFromPlugins(type, name, args);
+    hooks.forEach(function (hook) {
+      hook();
     });
-    return ran;
+
+    return !!hooks.length;
```
Comment (christophwitzko) [inline lib/core/hooks.js]: If there is at least one hook, return `true`.

File: lib/core/hooks.js [hunk 4]
```diff
@@ -3,30 +3,33 @@ var async = require('async');
...
   // runs a hook for all registered plugins
   var run = function (type, name, args) {
-
     if (env_config.verbose) {
       console.log('run %s hook: %s', type, name);
     }
 
-    var ran = false;
-
-    // for all plugins
```
Comment (christophwitzko) [inline lib/core/hooks.js]: The next lines are nearly the same as in the `getHooksFromPlugins` function.

**Comment by christophwitzko** [comment]
Hi @janl,

I have added some comments to the code. Hope it helps. :smile:

File: lib/core/hooks.js (hunk 1 above)
Comment (janl) [inline lib/core/hooks.js]: hmm, this might be our memory leak from https://github.com/hoodiehq/hoodie-server/issues/280!

**Comment by janl** [comment]
cool, thanks!

File: lib/core/hooks.js (hunk 1 above)
Comment (christophwitzko) [inline lib/core/hooks.js]: I guess you're right. :+1:

**Comment by christophwitzko** [comment]
Thanks. :relaxed:

File: lib/core/hooks.js (hunk 1 above)
Comment (inator) [inline lib/core/hooks.js]: While this very well may an issue, the versions I've been testing with have identical hooks.js files, yet the newer server version leaks.  :( I'll still run tests against the newer build for reference though and let you know.

**Pull request merged**